### PR TITLE
Define default collation for TiDB tables

### PR DESCRIPTION
**Pull request by mneudert**
### Description:

TiDB uses `utf8mb4_general_bin` as the default collation when creating a table. Compared to the MySQL default of `utf8mb4_general_ci` the sorting of some database results will be different.

This PR changes the table/database creation code to be more consistent across engines and actions:

1. Always define `ROW_FORMAT=dynamic`

   Previously, the row format was only defined for tables created during a migration.

   For InnoDB, all tables currently should default to `ROW_FORMAT=DYNAMIC`, adding it to all tables by default should (as far as I know) not be an issue.

   _Note_: TiDB currently ignores the `ROW_FORMAT` and always uses `COMPACT` ([docs](https://docs.pingcap.com/tidb/v8.1/information-schema-tables)).

2. Set a default collation for TiDB tables

   When TiDB is configured, and the (default) `utf8mb4` charset is used, all created tables will be created with a matching `COLLATE=utf8mb4_0900_ai_ci`.

4. Set a default collation for TiDB databases

   Creating a database also sets the default collation if TiDB is configured.

Refs DEV-18061

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: core/Db/Schema/Mysql.php [hunk 1]
```diff
@@ -680,16 +673,48 @@ public function getDefaultPort(): int
...
 
-    private function getTableEngine()
+    protected function getTableEngine()
     {
         return $this->getDbSettings()->getEngine();
     }
 
+    protected function getTableRowFormat(): string
+    {
+        return $this->getDbSettings()->getRowFormat();
```
Comment (sgiehl) [inline core/Db/Schema/Mysql.php:705]: Should we maybe fully move the `getRowFormat` away from `DbSettings` into the `Schema`? That way we could also simply leave it away for TiDb in general. Or if ever needed use another one for TiDb if they start adding something around that.

**Review by sgiehl — APPROVED** [review]
Left a comment for an improvement. But we could also change that as soon as it is needed though. Otherwise the PR looks good 👍

File: core/Db/Schema/Mysql.php [hunk 2]
```diff
@@ -522,10 +520,10 @@ public function createDatabase($dbName = null)
             $dbName = $this->getDbName();
         }
 
+        $createOptions = $this->getDatabaseCreateOptions();
```
Comment (michalkleiner) [inline core/Db/Schema/Mysql.php:523]: Just a minor suggestion — we're using `$tableOptions` var name above meaning table create options, here we use `$createOptions` var as db create options, might be good for consistency to always have the area it relates to (db/table) in the name in both places or in neither.

```suggestion
        $dbOptions = $this->getDatabaseCreateOptions();
```
or
```suggestion
        $dbCreateOptions = $this->getDatabaseCreateOptions();
```

**Review by michalkleiner — APPROVED** [review]

File: core/Db/Schema/Mysql.php (hunk 2 above)
Comment (michalkleiner) [reply core/Db/Schema/Mysql.php:523]: Never mind, I thought it was in the same method, but it's different methods, so no all good.

**Comment by michalkleiner** [comment]
> Left a comment for an improvement. But we could also change that as soon as it is needed though. Otherwise the PR looks good 👍

We can address this separately if needed, I'll merge this now so that we can run rebase and another set of tests for TiDB.

**Pull request merged**